pkg/querier/block.go

```diff
@@ -80,91 +152,208 @@ func (b *BlockQuerier) Get(ctx context.Context, userID string, from, through mod
 		}
 
-		it := enc.Iterator(nil)
-		for it.Next() {
-			ts, v := it.At()
-			overflow, err := ch.Add(model.SamplePair{
-				Timestamp: model.Time(ts),
-				Value:     model.SampleValue(v),
-			})
-			if err != nil {
-				return nil, errors.Wrap(err, fmt.Sprintf("failed adding sample to chunk (series: %v timestamp: %d value: %f)", lbls, ts, v))
-			}
+		it := ch.Iterator(nil)
+		its = append(its, it)
+	}
 
-			if overflow != nil {
-				chunks = append(chunks, chunk.NewChunk(userID, client.Fingerprint(lbls), lbls, ch, model.Time(c.MinTime), model.Time(c.MaxTime)))
-				ch = overflow
-			}
+	return newBlockQuerierSeriesIterator(bqs.Labels(), its)
+}
+
+func newBlockQuerierSeriesIterator(labels labels.Labels, its []chunkenc.Iterator) *blockQuerierSeriesIterator {
+	return &blockQuerierSeriesIterator{labels: labels, iterators: its, lastT: math.MinInt64}
+}
+
+// blockQuerierSeriesIterator implements a series iterator on top
+// of a list of time-sorted, non-overlapping chunks.
+type blockQuerierSeriesIterator struct {
+	// only used for error reporting
+	labels labels.Labels
+
+	iterators []chunkenc.Iterator
+	i         int
+	lastT     int64
+}
+
+func (it *blockQuerierSeriesIterator) Seek(t int64) bool {
+	// We generally expect the chunks already to be cut down
+	// to the range we are interested in. There's not much to be gained from
+	// hopping across chunks so we just call next until we reach t.
+	for {
+		ct, _ := it.At()
+		if ct >= t {
+			return true
+		}
+		if !it.Next() {
+			return false
 		}
+	}
+}
+
+func (it *blockQuerierSeriesIterator) At() (int64, float64) {
+	if it.i >= len(it.iterators) {
+		return 0, 0
+	}
+
+	t, v := it.iterators[it.i].At()
+	it.lastT = t
+	return t, v
+}
+
+func (it *blockQuerierSeriesIterator) Next() bool {
+	if it.i >= len(it.iterators) {
+		return false
+	}
+
+	if it.iterators[it.i].Next() {
+		return true
+	}
+	if it.iterators[it.i].Err() != nil {
+		return false
+	}
+
+	for {
+		it.i++
 
-		// Ensure the iteration has not been interrupted because of an error
-		if it.Err() != nil {
-			return nil, errors.Wrap(it.Err(), fmt.Sprintf("failed reading sample from encoded chunk (series: %v min time: %d max time: %d)", lbls, c.MinTime, c.MaxTime))
+		if it.i >= len(it.iterators) {
+			return false
 		}
 
-		if ch.Len() > 0 {
-			chunks = append(chunks, chunk.NewChunk(userID, client.Fingerprint(lbls), lbls, ch, model.Time(c.MinTime), model.Time(c.MaxTime)))
+		// we must advance iterator first, to see if it has any samples.
+		// Seek will call At() as its first operation.
+		if !it.iterators[it.i].Next() {
+			if it.iterators[it.i].Err() != nil {
+				return false
+			}
+
+			// Found empty iterator without error, skip it.
+			continue
 		}
+
+		// Chunks are guaranteed to be ordered but not generally guaranteed to not overlap.
+		// We must ensure to skip any overlapping range between adjacent chunks.
+		return it.Seek(it.lastT + 1)
```

Comment:
If `Seek()` return `false` we return `false` from `Next()` too. Shouldn't we skip to the next iterator?

Reply:
`Seek` will skip to next iterator, since it calls `Next`, which will do it. There is a bit of recursion here, if you need to skip over several empty iterators.

Reply:
You're right. My bad.